src/Dsn.php

```diff
@@ -76,32 +76,32 @@ private function __construct(string $scheme, string $host, int $port, int $proje
      */
     public static function createFromString(string $value): self
     {
+        /** @var array{scheme:string, host:string, path:string, user:string, pass?:string}|false */
```

Comment:
Where are it DSN parts defined? I think nowhere.

Reply:
If we do not declare it then `@psalm-suppress PossiblyUndefinedArrayOffset` AND PHPStan ignores could not be removes.

That is the sole point of the PR!

Reply:
>Where are DSN parts defined? I think nowhere.

PHPStan is able to infer the keys of the array from the stubs of the `parse_url` function

>If we do not declare it then `@psalm-suppress PossiblyUndefinedArrayOffset` and PHPStan ignores could not be removed.

In reality, I believe that this error is just a bug in PHPStan that is not handling the `empty($parsedDsn[$component])` check inside the `foreach` properly. Instead of declaring again the shape of the array I would try to use `isset()` and see if it solves the issue

Reply:
Please see https://github.com/phpstan/phpstan/issues/3783#issuecomment-680304162

Reply:
`parse_url` definition in PHPStan has all the keys optional because it cannot presume what's in the input value (https://phpstan.org/r/cc7650f5-e052-4f0e-be1e-e8973750fc93), any part of the URL can be missing in it. Maybe that's why @szepeviktor added the `@var`? Because it's known the string looks a certain way?

Reply:
>Because it's known the string looks a certain way?

We know the string should have all those components, but of course we cannot be sure of this and it's the reason there is a loop that does a `if (!empty($parsedDsn[$component])) { throw ... }`. However, the `empty` function is not being handled as if it was an `isset` (at least in this case): https://phpstan.org/r/61073b70-73ea-41f9-b3a6-31c68291e80f

Reply:
Actually those elements are checked with `empty()` in a for loop below in line number 86.

Reply:
Interesting, so `!isset` works: https://phpstan.org/r/c102939c-132d-44c9-a159-f93b35d83ac1

But `empty` doesn't: https://phpstan.org/r/26cf5f07-d39c-4d81-bbd8-80850dc57411

Can you open an issue in phpstan/phpstan about that? Thanks.

Reply:
Done: phpstan/phpstan#3784